apps/prairielearn/src/lib/groups.sql

```diff
@@ -172,6 +128,122 @@ WHERE
   AND gu.user_id = $user_id
   AND g.deleted_at IS NULL;
 
+-- BLOCK select_group_by_name
+SELECT
+  g.id
+FROM
+  groups AS g
+  JOIN group_configs AS gc ON g.group_config_id = gc.id
+WHERE
+  g.name = $group_name
+  AND g.join_code = $join_code
+  AND gc.assessment_id = $assessment_id
+  AND g.deleted_at IS NULL
+  AND gc.deleted_at IS NULL
+FOR NO KEY UPDATE of
+  g;
+
+-- BLOCK select_group_for_update
+SELECT
+  g.*,
+  (
+    SELECT
+      COUNT(*)
+    FROM
+      group_users AS gu
+    WHERE
+      gu.group_id = g.id
+  )::INT AS cur_size,
+  gc.maximum AS max_size,
+  gc.has_roles
+FROM
+  groups AS g
+  JOIN group_configs AS gc ON g.group_config_id = gc.id
+WHERE
+  g.id = $group_id
+  AND gc.assessment_id = $assessment_id
+  AND g.deleted_at IS NULL
+  AND gc.deleted_at IS NULL
+FOR NO KEY UPDATE of
+  g;
+
+-- BLOCK select_suitable_group_role
+WITH
+  users_per_group_role AS (
+    SELECT
+      gur.group_role_id,
+      COUNT(*) AS user_count
+    FROM
+      group_user_roles AS gur
+    WHERE
+      gur.group_id = $group_id
+    GROUP BY
+      gur.group_role_id
+  ),
+  suitable_group_roles AS (
+    SELECT
+      gr.id,
+      gr.can_assign_roles_at_start
+      AND COALESCE($cur_size::INT, 0) = 0 AS assigner_role_needed,
+      upgr.user_count IS NULL
+      AND gr.minimum > 0 AS mandatory_with_no_user,
+      GREATEST(0, gr.minimum - COALESCE(upgr.user_count, 0)) AS needed_users,
+      gr.maximum - COALESCE(upgr.user_count, 0) AS remaining
+    FROM
+      group_roles AS gr
+      LEFT JOIN users_per_group_role AS upgr ON (upgr.group_role_id = gr.id)
+    WHERE
+      gr.assessment_id = $assessment_id
+      AND (gr.maximum IS NULL OR gr.maximum > COALESCE(upgr.user_count, 0))
+  )
+SELECT
+  sgr.id
+FROM
+  suitable_group_roles AS sgr
+ORDER BY
+  -- If there are no users in the group, assign to an assigner role.
+  assigner_role_needed DESC,
+  -- Assign to a mandatory role with no users, if one exists.
+  mandatory_with_no_user DESC,
+  -- Assign to role that has not reached their minimum.
+  needed_users DESC,
+  -- Assign to role that has the most remaining spots.
+  remaining DESC NULLS LAST
+limit
+  1;
```

Comment:
This is a slight change in behaviour, and I'd like to get your thoughts on this. Assume a scenario where there are 4 roles: Manager, Recorder, Reflector, Contributor. The first three roles have a minimum and maximum of 1, the last one has no minimum or maximum. This is the basic role scenario used in the example course and in the tests.

The first user in the group (typically the one that creates it) will be assigned the Manager role (the role that permits assigning other roles). This remains the same (via the "assigner_role_needed" check above).

The following two students in the group are where the change in behaviour happens. Currently they are assigned to some role that has a minimum larger than zero and no assigner role, but there is no control over making sure these roles are different. In particular, in the example above this would assign the next two students to the Recorder role, which would violate the maximum numbers for Recorder and would keep the Reflector role empty. The new behaviour is that the students would be assigned to the first role that still needs a student, with priority to roles that have no student yet. In the example above, the next two students would thus be assigned one for Recorder and one for Reflector.

What baffles me in the previous process is that the decision to move on with optional roles (Contributor in the case above) is based on the number of students in the group in general, even if there are roles still to be filled with their minimum counts. In the new behaviour, roles with no minimum numbers would only be assigned if there are no more roles where the minimum is not met.

Now the question: the tests rely on the behaviour above, which to me is confusing:
https://github.com/PrairieLearn/PrairieLearn/blob/b2726d342307130f1dfe484362cef102b109a42a/apps/prairielearn/src/tests/groupRole.test.js#L522-L528
Is this indeed the expected behaviour, or were the tests just created this way because it was the observed behaviour? I can for sure change the tests, I just want to make sure that there is no intentional reason to have the old behaviour (the "two recorder students with no reflector student") be the default were even tests are checking for this to be the case?

@mwest1066 @nwalters512 @renzol2

Reply:
Discussed in our meeting. The new role assignment process is better, and tests should account for that.